rust/sbp/src/sbp_string.rs

```diff
@@ -37,6 +39,30 @@ where
     }
 }
 
+impl<T: AsRef<[u8]>> SbpString<T, Multipart> {
+    pub fn parse(&self) -> Result<Vec<std::borrow::Cow<str>>, MultipartError> {
+        parse_to_vec(self.as_bytes(), 1)
+    }
+}
+
+impl<T: AsRef<[u8]>> SbpString<T, NullTerminated> {
+    pub fn parse(&self) -> Result<std::borrow::Cow<str>, MultipartError> {
```

Comment:
I think this should be 


```suggestion
    pub fn parts(&self) -> impl Iterator<Item=&[u8]> {
```

It's more flexible (doesn't allocate and doesn't parse as a string) but you can replicate the current behavior with it. Also I don't think Result should be involved here, you should assume that the string is a valid multipart string (and then provide constructors for multipart string that makes sure they are valid). I was thinking of adding a constructor for each encoding on SbpString so like SbpString::multipart which would be fallible

Reply:
Something like:

```rust
impl<T: AsRef<[u8]>, E> SbpString<T, E> {
    pub fn parts(&self) -> impl Iterator<Item = &[u8]> {
        self.as_bytes().split(|b| b == &0)
    }
}
```

and have 

```rust
pub struct PartsDecoder<T>(T);

impl PartsDecoder<DoubleNullTerminated>{
    
    fn parse(parts: impl Iterator<Item = &[u8]>) -> Result<Vec<Cow<str>>, MultipartError>{
        parts_to_vec(parts, 2)
    }
    
}
```
and have `parts_to_vec` as a helper method to parse iterator according to the nulls, return an error if the number of nulls dont match?

Reply:
I was thinking the API could look something like 

```rust
/// Fixed or variable length string and its encoding.
#[derive(Debug, PartialEq, Clone)]
pub struct SbpString<T, E = Unterminated> {
    data: T,
    encoding: PhantomData<E>,
}

impl<T, E> SbpString<T, E>
where
    T: AsRef<[u8]>,
{
    /// Creates a new SbpString with a given encoding. Does not check if the data is valid
    /// for the given encoding.
    pub fn new_unchecked(data: T) -> Self { // rename new to reflect that it doesn't check encodings
        Self {
            data,
            encoding: PhantomData,
        }
    }

    /// Returns a byte slice of this SbpString's contents.
    pub fn as_bytes(&self) -> &[u8] {
        self.data.as_ref()
    }

    /// Returns a byte vector of this SbpString's contents.
    pub fn to_vec(&self) -> Vec<u8> {
        self.data.as_ref().to_vec()
    }
}

impl<T: AsRef<[u8]>> SbpString<T, Unterminated> {
    // I think this one can be infallible
    pub fn unterminated(data: T) -> Self {
        todo!();
    }
}

impl<T: AsRef<[u8]>> SbpString<T, NullTerminated> {
     // maybe doesn't need to return a result we could also just append the null byte if missing
    pub fn null_terminated(data: T) -> Result<Self, NullTerminatorError> {
        todo!();
    }
}

// from calling .split() on a slice/vec
// we should reuse this iterator because it impls a lot of nice things
// like DoubleEndedIterator. And its a std type so it's fine to have it be in the public API
// We could also wrap it up in a newtype if you want
pub type Parts<'a> = std::slice::Split<'a, u8, fn(&u8) -> bool>;

// Multipart and DoubleNullTerminated all variable length so we don't need to
// be generic over the container we can just assume Vec<u8>

impl SbpString<Vec<u8>, Multipart> {
    pub fn multipart(data: impl Into<Vec<u8>>) -> Result<Self, MultipartError> {
        todo!();
    }

    pub fn from_parts(parts: impl IntoIterator<Item = impl AsRef<[u8]>>) -> Self {
        todo!();
    }

    pub fn parts(&self) -> Parts<'_> {
        todo!();
    }
}

impl SbpString<Vec<u8>, DoubleNullTerminated> {
    pub fn double_null_terminated(data: impl Into<Vec<u8>>) -> Result<Self, DoubleNullTerminatedError> {
        todo!();
    }

    pub fn from_parts(parts: impl IntoIterator<Item = impl AsRef<[u8]>>) -> Self {
        todo!();
    }

    // cant share an impl with Multipart bc you need to handle the extra null at the end
    pub fn parts(&self) -> Parts<'_> {
        todo!();
    }
}
```

The one thing I don't like about this API is that if you wanna build a new string with `from_parts` I think you'll still need to specify the encoding.

```rust
// is this multipart or double null terminated?
let x = SbpString::from_parts(["a", "b", "c"]);
// will have to do something like
let x = SbpString::<_, Multipart>::from_parts(["a", "b", "c"]);
```

Not sure if there is a nice way around that, other than by naming the functions differently. Also something to consider is adding `FromIterator` impls for the multipart strings

```rust
impl<T: AsRef<[u8]>> FromIterator<T> for SbpString<Vec<u8>, Multipart> {
    fn from_iter<I: IntoIterator<Item=T>>(iter: I) -> Self {
        todo!();
    }
}
```

which would let people use `.collect` to create multipart strings

```rust
let parts = ["a", "b", "c"];
let string: SbpString<_, Multipart> = parts.into_iter().collect();
```

Could just do that instead of `from_parts` (or could do both and impl one via the other)

> and have parts_to_vec as a helper method to parse iterator according to the nulls, return an error if the number of nulls dont match?

Creating the string from the parts should be something like 

```rust
fn from_parts(it: ...) -> Self {
    // check that none of the parts contain a null byte already
    // and return an error (I think a case could also be made to panic instead, or we could have the code
    // remove the null bytes) 
    // once you've checked it just
    Self {
        data: it.intersperse(b"0").collect(),
        encoding: PhantomData,
    }
}
```